lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp

```diff
@@ -611,6 +614,121 @@ void CanonicalizeOSSALifetime::insertDestroysOnBoundary(
   }
 }
 
+/// At -Onone, there are some conflicting goals:
+/// On the one hand: good debugging experience.
+/// (1) do not shorten value's lifetime
+/// On the other: demonstrate semantics.
+/// (2) consume value at same places it will be consumed at -O
+/// (3) ensure there are no more copies than there would be at -O
+///
+/// (2) and (3) are equivalent--extra (compared to -O) copies arise from failing
+/// to end lifetimes at consuming uses (which then need their own copies).
+///
+/// We achieve (2) and (3) always.  We achieve (1) where possible.
+///
+/// Conceptually, the strategy is the following:
+/// - Collect the blocks in which the value was live before canonicalization.
+///   These are the "original" live blocks (originalLiveBlocks).
+///   [Color these blocks green.]
+/// - From that collection, collect the blocks which occur after a _final_
+///   consuming use (that isn't a destroy).
+///   These are the "consumed" blocks (consumedBlocks).
+///   [Color these blocks red.]
+/// - Extend liveness up to the boundary between originalLiveBlocks and
+///   consumedBlocks blocks.
+///   [Extend liveness up to the boundary between green blocks and red.]
+/// - In particular, in regions of originalLiveBlocks which have no boundary
+///   with consumedBlocks, livness should be extended to its original extent.
+///   [Extend liveness up to the boundary between green blocks and uncolored.]
+void CanonicalizeOSSALifetime::extendUnconsumedLiveness(
+    PrunedLivenessBoundary &boundary) {
+  auto currentDef = getCurrentDef();
+
+  // First, collect the blocks that were _originally_ live.  We can't use
+  // liveness here because it doesn't include blocks that occur before a
+  // destroy_value.
+  BasicBlockSet originalLiveBlocks(currentDef->getFunction());
+  {
+    // Start the walk from the consuming blocks (which includes destroys as well
+    // as the other consuming uses).
+    BasicBlockWorklist worklist(currentDef->getFunction());
+    for (auto *consumingBlock : consumingBlocks) {
+      worklist.push(consumingBlock);
+    }
+
+    // Walk backwards from consuming blocks.
+    auto *currentDefBlock = currentDef->getParentBlock();
+    while (auto *block = worklist.pop()) {
+      originalLiveBlocks.insert(block);
```

Comment:
So `liveness.liveBlocks` is already a potentially large subset of `originalLiveBlocks`. computeCanonicalLiveness already did that work for you. A block only needs to be pushed onto the "original live" worklist if it's not originally in liveBlocks. And there's no need to check for the def block. You might need a wrapper to check for membership in both places... but it seems worthwhile.

Reply:
That's great, thanks!  Just adding the blocks discovered by liveness to the set here--that way we only iterate the array once.  (Also, more readable than checking membership in two places.)